packages/suite/src/middlewares/wallet/coinjoinMiddleware.ts

```diff
@@ -55,6 +62,35 @@ export const coinjoinMiddleware =
             api.dispatch(coinjoinAccountActions.pauseCoinjoinSessionByDeviceId(action.payload.id));
         }
 
+        if (action.type === COINJOIN.CLIENT_SESSION_PHASE) {
+            const account = api
+                .getState()
+                .wallet.coinjoin.accounts.find(({ key }) => key === action.payload.accountKeys[0]); // all accounts asocciated with the accountKeys array have the same session phase
+
+            if (account) {
+                const { key, session } = account;
+
+                const queueLength = session?.sessionPhaseQueue?.length;
+                const preveiousState = previousAccountsState
+                    .find(account => account.key === key)
+                    ?.session?.sessionPhaseQueue.at(-1);
+
+                const isSamePhase = action.payload.phase === preveiousState;
+
+                if (queueLength && queueLength > 1 && !isSamePhase) {
+                    const nextPhaseDelay =
+                        SESSION_PHASE_TRANSITION_DELAY -
+                        (Date.now() - (session?.lastSessionPhaseChangeTimestamp || Date.now()));
+                    const otherPhasesDelay = SESSION_PHASE_TRANSITION_DELAY * (queueLength - 2);
+                    const timeToDelay = nextPhaseDelay + otherPhasesDelay;
+
+                    setTimeout(() => {
+                        api.dispatch(clientShiftSessionPhase(action.payload.accountKeys));
```

Comment:
what if another COINJOIN.CLIENT_SESSION_PHASE action happen before this timeout expires? Is there no need to clear this timeout?

Reply:
That's the point, another timeout is scheduled with a larger delay :)